tenacity/asyncio/retry.py

```diff
@@ -104,7 +104,7 @@ def __init__(self, *retries: typing.Union[retry_base, async_retry_base]) -> None
     async def __call__(self, retry_state: "RetryCallState") -> bool:  # type: ignore[override]
         result = False
         for r in self.retries:
-            result = result or await _utils.wrap_to_async_func(r)(retry_state)
+            result = result or (await _utils.wrap_to_async_func(r)(retry_state) is True)
```

Comment:
I see you added `is True` everywhere. This seems like a different change not required. Are we sure we need this?

Reply:
While adding the tests I noticed that, if a wrong combination of strategies is added (combination of async strategies in a sync context), then we'd end up trying to `and`/`or` a coroutine, which hangs forever as it's never going to be resolved in the sync context. By adding the `is True` we directly check if the result is exactly that and avoid endlessly waiting for the coroutine.

Reply:
Actually, thinking about it a bit more... Maybe we'd rather let it hang? Not sure what's best they, without the changes it hangs forever, which is not very intuitive, and with these changes it finishes without failures but it does not run the strategy at all, giving an invalid outcome 😕

Reply:
Is it feasible to trap this invalid call pattern and raise an exception to make it obvious that there's a programming error?

Reply:
I came up with this, it is more coupled to the actual implementations, but it now fails on invalid scenarios so it's more explicit. I used a similar check to [the one already existing in `__init__.py`](https://github.com/jd/tenacity/blob/31fe2d0cf2505bffd0cf1ffda8c7e30450ce709f/tenacity/__init__.py#L643-L650), let me know what you think @jd https://github.com/jd/tenacity/pull/485/commits/f8725d6abbfcaf38eb36725c31bf7fe80fee1ab0